asm-utils/src/data_storage.rs

```diff
@@ -2,76 +2,157 @@
+use crate::{
+    data_parser::DataValue,
+    utils::{alignment_size, next_aligned},
+};
 
 pub enum SingleDataValue<'a> {
     Value(u32),
-    LabelReference(&'a String),
-    Offset(&'a String, &'a String),
+    LabelReference(&'a str),
+    Offset(&'a str, &'a str),
 }
 
-pub fn store_data_objects<'a>(
-    objects: impl IntoIterator<Item = &'a (String, Vec<DataValue>)> + Copy,
+struct WordWriter<'a, 'b> {
+    data_writer: &'a mut dyn FnMut(u32, SingleDataValue) -> Vec<String>,
+    partial: u32,
+    current_pos: u32,
+    generated_code: Vec<String>,
+
+    latest_label: Option<&'b str>,
+}
+
+impl<'a, 'b> WordWriter<'a, 'b> {
+    fn new(
+        starting_pos: u32,
+        data_writer: &'a mut dyn FnMut(u32, SingleDataValue) -> Vec<String>,
+    ) -> Self {
+        // sanitary alignment to 8 bytes
+        let current_pos = next_aligned(starting_pos as usize, 8) as u32;
+        Self {
+            partial: 0,
+            current_pos,
+            data_writer,
+            generated_code: Vec::new(),
+            latest_label: None,
+        }
+    }
+
+    fn current_position(&self) -> u32 {
+        self.current_pos
+    }
+
+    fn set_label(&mut self, label: &'b str) {
+        self.latest_label = Some(label)
+    }
+
+    fn advance(&mut self, bytes: u32) {
+        let next_pos = self.current_pos + bytes;
+
+        // if changed words, flush
+        let curr_word = self.current_pos & (!0b11);
+        if (next_pos & (!0b11) != curr_word) && (self.partial != 0) {
+            if let Some(label) = std::mem::take(&mut self.latest_label) {
+                self.generated_code.push(format!("// data {label}"));
+            }
+
+            self.generated_code.extend((*self.data_writer)(
+                curr_word,
+                SingleDataValue::Value(self.partial),
+            ));
+            self.partial = 0;
+        }
+        self.current_pos = next_pos;
+    }
+
+    fn align(&mut self, alignment: u32) {
+        let padding_size = alignment_size(self.current_pos as usize, alignment as usize);
+        if padding_size != 0 {
+            self.advance(padding_size as u32);
+        }
+    }
+
+    fn write_bytes(&mut self, bytes: &[u8]) {
+        for b in bytes {
+            self.partial |= (*b as u32) << (8 * (self.current_pos % 4));
+            self.advance(1);
+        }
+    }
+
+    fn write_label_reference(&mut self, label: &str) {
+        assert_eq!(
+            self.current_pos % 4,
+            0,
+            "reference to code labels in misaligned data section is not supported"
+        );
+
+        self.generated_code.extend((*self.data_writer)(
+            self.current_pos,
+            SingleDataValue::LabelReference(label),
+        ));
+
+        assert_eq!(self.partial, 0);
+        self.current_pos += 4;
+    }
+
+    fn finish(mut self) -> Vec<String> {
+        // ensure the latest partial word is written
+        self.advance(4);
```

Comment:
why 4?

Reply:
Because that is the size of a word.

Reply:
that's for RISCV-32 though, maybe this library should be parametric over the native word size

Reply:
This is more about powdr's word than RV32. And I don't see powdr's memory word being larger that 32 bits, otherwise it would break Goldilocks.

That said, making this struct generic is easy enough, but probably would depend on `num_traits` crate. There was another thing that I did not made generic over native ints because `asm-utils` did not on `num_traits`.

Reply:
That's fine, the only thing I think doesn't fit is that this is supposed to be "asm-utils" which would also work for WASM for example. But we can also go for WASM32 and everything still works. Fine to keep.

Reply:
Even for 64-bit processors, the memory word in the zkvm can't be bigger than the smallest field we support, so we would still use 4 byte words.